src/appdistribution/options-parser-util.ts

```diff
@@ -62,3 +64,109 @@ export function getAppName(options: any): string {
   const appId = options.app;
   return `projects/${appId.split(":")[1]}/apps/${appId}`;
 }
+
+/**
+ * Takes in comma separated string or a path to a comma/new line separated file
+ * and converts the input into a string[] of test device strings. Value takes precedent
+ * over file.
+ */
+export function getTestDevices(value: string, file: string): TestDevice[] {
+  // If there is no value then the file gets parsed into a string to be split
+  if (!value && file) {
+    ensureFileExists(file);
+    value = fs.readFileSync(file, "utf8");
+  }
+
+  if (!value) {
+    return [];
+  }
+
+  return value
+    .split(/[;\n]/)
+    .map((entry) => entry.trim())
+    .filter((entry) => !!entry)
+    .map((str) => parseTestDevice(str));
+}
+
+function parseTestDevice(testDeviceString: string): TestDevice {
+  const entries = testDeviceString.split(",");
+  const allowedKeys = new Set(["model", "version", "orientation", "locale"]);
+  let model: string | undefined;
+  let version: string | undefined;
+  let orientation: string | undefined;
+  let locale: string | undefined;
+  for (const entry of entries) {
+    const keyAndValue = entry.split("=");
+    switch (keyAndValue[0]) {
+      case "model":
+        model = keyAndValue[1];
+        break;
+      case "version":
+        version = keyAndValue[1];
+        break;
+      case "orientation":
+        orientation = keyAndValue[1];
+        break;
+      case "locale":
+        locale = keyAndValue[1];
+        break;
+      default:
+        throw new FirebaseError(
+          `Unrecognized key in test devices. Can only contain ${Array.from(allowedKeys).join(", ")}`,
+        );
+    }
+  }
+
+  if (!model || !version || !orientation || !locale) {
+    throw new FirebaseError(
+      "Test devices must be in the format 'model=<model-id>,version=<os-version-id>,locale=<locale>,orientation=<orientation>'",
+    );
+  }
+  return { model, version, locale, orientation };
+}
+
+/**
+ * Takes option values for username and password related options and returns a LoginCredential
+ * object that can be passed to the API.
+ */
+export function getLoginCredential(
+  username?: string,
```

Comment:
Optional: This would be a good spot to use a parameter object - with 5 optional fields all with the same type, it seems very easy to accidentally pass in args in the wrong order. A param object makes this a bit more explicit and readable:
```
function getLoginCredential( args: {
  username?: string,
  password?: string,
  passwordFile?: string,
  usernameResourceName?: string,
  passwordResourceName?: string,
})

Reply:
I like it, done